packages/workspace/src/browser/workspace-trust-service.ts

```diff
@@ -0,0 +1,101 @@
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { Dialog, StorageService } from '@theia/core/lib/browser';
+import { PreferenceService } from '@theia/core/lib/browser/preferences/preference-service';
+import { MessageService } from '@theia/core/lib/common/message-service';
+import { nls } from '@theia/core/lib/common/nls';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { inject, injectable, postConstruct } from '@theia/core/shared/inversify';
+import { WorkspaceService } from '.';
+import {
+    WorkspaceTrustPreferences, WORKSPACE_TRUST_EMPTY_WINDOW, WORKSPACE_TRUST_ENABLED, WORKSPACE_TRUST_STARTUP_PROMPT, WORKSPACE_TRUST_STARTUP_PROMPT_NEVER,
+    WORKSPACE_TRUST_STARTUP_PROMPT_ONCE
+} from './workspace-trust-preferences';
+
+const STORAGE_TRUSTED = 'trusted';
+
+@injectable()
+export class WorkspaceTrustService {
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @inject(PreferenceService)
+    protected readonly preferences: PreferenceService;
+
+    @inject(StorageService)
+    protected readonly storage: StorageService;
+
+    @inject(MessageService)
+    protected readonly messageService: MessageService;
+
+    @inject(WorkspaceTrustPreferences)
+    protected readonly workspaceTrustPref: WorkspaceTrustPreferences;
+
+    protected trusted = new Deferred<boolean>();
+
+    @postConstruct()
+    protected async init(): Promise<void> {
+        await this.workspaceService.ready;
+        const implicitTrust = await this.calculateWorkspaceTrust();
+        if (implicitTrust !== undefined) {
+            this.trusted.resolve(implicitTrust);
+        }
+    }
+
+    getWorkspaceTrust(): Promise<boolean> {
+        return this.trusted.promise;
+    }
+
+    protected async calculateWorkspaceTrust(): Promise<boolean | undefined> {
+        if (!this.workspaceTrustPref[WORKSPACE_TRUST_ENABLED]) {
+            // in VS Code if workspace trust is disabled, we implicitly trust the workspace
+            return true;
+        }
+
+        if (this.workspaceTrustPref[WORKSPACE_TRUST_EMPTY_WINDOW] && !this.workspaceService.workspace) {
+            return true;
+        }
+
+        const trustPrompt = this.workspaceTrustPref[WORKSPACE_TRUST_STARTUP_PROMPT];
+        if (trustPrompt === WORKSPACE_TRUST_STARTUP_PROMPT_ONCE) {
+            return this.storage.getData<boolean>(STORAGE_TRUSTED);
+        }
+
+        // reset stored trusted setting
+        this.storage.setData(STORAGE_TRUSTED, undefined);
```

Comment:
The handling of the storage is a bit odd. Basically, we only care about storing the value if the user has set `PROMPT_ONCE` - and we want to be able to save the value if the user changes that preference value after we run `requestWorkspaceTrust`, so perhaps the saving / deleting should be handled in a preference change listener, as well?

Reply:
I will implement that in a follow-up change as well, I think that might be a good idea, thank you!